### PR TITLE
Feature/pack fuse

### PR DESCRIPTION
**Pull request by maddyscientist**
Significant clean up of the dslash kernels, some bug fixes, and adds some new Dslash policies
* Two new dslash policies - fused Dslash + packing kernel policies (fused and unfused exterior variants), where the face packing and interior dslash application are done in the same kernel.
* Cleanup of dslash kernels to use a common generic kernel `dslashGPU` for all dslash types with operators defined using a functor in a class
* Filenames are now automatically deducted with `filename` helper function in operator classes
* Fix multi-blas compilation on Jitify - erroneous use of `std::max` on list which isn't supported in Jitify yet
* Jitify compilation would often fail due to flaky object resolution of parent `Dslash` class methods - this has now been made robust by templating the `Dslash` parent class on the operator class type (as a template template parameter) and the kernel argument struct type.  This also facilitates a significant cleanup of the driver code, with much more code moved to the parent `Dslash`, requiring only minimal code in the derived class.

### CODE REVIEW

File: include/kernels/dslash_staggered.cuh [hunk 1]
```diff
@@ -6,17 +6,19 @@
 #include <color_spinor.h>
 #include <dslash_helper.cuh>
 #include <index_helper.cuh>
+#include <kernels/laplace.cuh>
```
Comment (mathiaswagner) [inline include/kernels/dslash_staggered.cuh]: Not quite obvious why this is included here

File: include/dslash.h [hunk 2]
```diff
@@ -209,38 +282,48 @@ namespace quda
...
-      const auto kernel = Launch<void, 0, 0, 0, false, false, INTERIOR_KERNEL, Arg>::kernel;
-      Tunable::jitify_error = program_->kernel(kernel)
-                                  .instantiate(Type<Float>(), nDim, nColor, arg.nParity, arg.dagger, arg.xpay,
-                                      arg.kernel_type, Type<Arg>())
-                                  .configure(tp.grid, tp.block, tp.shared_bytes, stream)
-                                  .launch(arg);
+      const auto kernel = "quda::dslashGPU";
+
+      // we need this hackery to get the naked unbound template class parameters
+      auto D_instance = reflect<D<void, 0, 0, 0, false, false, INTERIOR_KERNEL, Arg>>();
```
[the remark this answers is not in the document]
Comment (maddyscientist) [reply include/dslash.h]: Done and pushed in 8cca57a

File: include/dslash.h [hunk 3]
```diff
@@ -9,25 +9,23 @@
 namespace quda
 {
 
-  template <typename Float> class Dslash : public TunableVectorYZ
+  template <template <typename, int, int, int, bool, bool, KernelType, typename> class D, typename Float, typename Arg>
+  class Dslash : public TunableVectorYZ
```
[the remark this answers is not in the document]
Comment (maddyscientist) [reply include/dslash.h]: Done

File: include/kernels/dslash_staggered.cuh (hunk 1 above)
Comment (maddyscientist) [reply include/kernels/dslash_staggered.cuh]: Should have been dslash_pack.cuh, fixed.

**Comment by mathiaswagner** [comment]
The new policies for fused pack are not yet covered in the separate policy ctest runs. We should add them there.

**Comment by maddyscientist** [comment]
Yep, that’s in commit I haven’t pushed yet.  Will be in next push from me later today.

**Pull request merged**